### PR TITLE
Transaction options part of RFC1004

### PR DESCRIPTION
**Pull request by tailhook**
This is the JavaScript counterpart of edgedb/edgedb-python#174

### CODE REVIEW

File: src/options.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
+import * as errors from "./errors";
+
+export type BackoffFunction = (n: number) => number;
+
+export function default_backoff(attempt: number): number {
```
Comment (jaclarke) [inline src/options.ts:5]:
```suggestion
export function defaultBackoff(attempt: number): number {
```

File: src/client.ts [hunk 2]
```diff
@@ -229,94 +262,160 @@ class StandaloneConnection implements Connection {
...
+    config: NormalizedConnectConfig
+  ): Promise<S> {
+    const conn = new this(config);
+    await conn[INNER].reconnect();
+    return conn;
+  }
+}
+
+export class InnerConnection {
+  borrowed_for?: BorrowReason;
```
Comment (jaclarke) [inline src/client.ts:396]:
```suggestion
  borrowedFor?: BorrowReason;
```

File: src/client.ts [hunk 3]
```diff
@@ -88,11 +89,6 @@ function default_backoff(attempt: number): number {
   return 2 ** attempt * 100 + Math.random() * 100;
```
Comment (jaclarke) [inline src/client.ts:89]: I think this is redundant now? Looks like it's been moved to options.ts

File: src/options.ts (hunk 1 above)
Comment (tailhook) [reply src/options.ts]: Fixed, thanks!

File: src/client.ts (hunk 2 above)
Comment (tailhook) [reply src/client.ts]: Fixed. Thanks! (always mess up with camel-case convention :( )

File: src/client.ts (hunk 3 above)
Comment (tailhook) [reply src/client.ts]: Yes. Removed.

**Comment by tailhook** [comment]
Updated:
1. No partial options objects
2. `Conflict` condition instead of `Deadlock` / `SerializationError`
3. More tests

**Review by 1st1 — APPROVED** [review]

**Comment by tailhook** [comment]
Waits for #100

**Comment by 1st1** [comment]
Paul, can you rebase this PR and merge it?

**Pull request merged**